test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -167,54 +207,314 @@
+          (str "did not get back the same instance from the third borrow"
+               "attempt as was returned from the first borrow attempt")))))
+
+(deftest pool-can-borrow-after-borrow-timed-out-during-lock
+  (testing (str "can do a timed borrow from pool after previous borrow "
+                "timed out while a write lock was held")
+    (let [pool (create-populated-pool 1)
+          borrow-1 (.borrowItem pool)
+          borrow-thread-started-2? (promise)
+          borrow-thread-started-3? (promise)
+          borrow-thread-2 (future
+                           (deliver borrow-thread-started-2? true)
+                           (.borrowItemWithTimeout
+                            pool
+                            50
+                            TimeUnit/MILLISECONDS))
+          borrow-thread-3 (future
+                           (deliver borrow-thread-started-3? true)
+                           (.borrowItemWithTimeout pool
+                                                   1
+                                                   TimeUnit/SECONDS))
+          _ @borrow-thread-started-2?
+          _ @borrow-thread-started-3?
+          unlock? (promise)
+          lock-thread-started? (promise)
+          lock-thread-locked? (promise)
+          lock-thread (future
+                       (deliver lock-thread-started? true)
+                       (.lock pool)
+                       (deliver lock-thread-locked? true)
+                       @unlock?
+                       (.unlock pool))]
+      @lock-thread-started?
+      (.releaseItem pool borrow-1)
+      @lock-thread-locked?
+      (is (true? true))
+      (is (nil? @borrow-thread-2)
+          "second borrow attempt did not time out")
+      (deliver unlock? true)
+      @lock-thread
+      (is (true? true))
+      (is (identical? borrow-1 @borrow-thread-3)
+          (str "did not get back the same instance from the third borrow"
+               "attempt as was returned from the first borrow attempt")))))
+
+(deftest pool-insert-pill-test
+  (testing "inserted pill is next item borrowed"
+    (let [pool (create-populated-pool 1)
+          pill (str "i'm a pill")]
+      (.insertPill pool pill)
+      (is (identical? (.borrowItem pool) pill)))))
+
+(deftest pool-clear-test
+  (testing (str "pool clear removes all elements from queue and only matching"
+                "registered elements")
+    (let [pool (create-populated-pool 3)
+          instance (.borrowItem pool)]
+      (is (= 2 (.size pool)))
+      (is (= 3 (-> pool
+                   (.getRegisteredElements)
+                   (.size))))
+      (.clear pool)
+      (is (= 0 (.size pool)))
+      (let [registered-elements (.getRegisteredElements pool)]
+        (is (= 1 (.size registered-elements)))
+        (is (identical? instance (-> registered-elements
+                                     (.iterator)
+                                     iterator-seq
+                                     first)))))))
```

Comment:
> We don't actually use clear in any production code anyway, AFAICT. 

Well, turns out I was wrong about that.  :disappointed:  There are a couple of places where we use it in production code:

https://github.com/puppetlabs/puppet-server/blob/puppet-server-2.2.1/src/clj/puppetlabs/services/jruby/jruby_puppet_agents.clj#L59

https://github.com/puppetlabs/puppet-server/blob/puppet-server-2.2.1/src/clj/puppetlabs/services/jruby/jruby_puppet_agents.clj#L115

In both cases, it is called after an exception is caught when a pool instance fails to be initialized.

I think calling `.clear` in these cases is wrong.  If we were to just do this without calling `.terminate` on any of the ScriptingContainers for the instances that had been created prior to the one that had failed, we would be leaking resources.

I'm thinking it might be best to just file a separate ticket for this one where we'd look into eliminating the `.clear` method and ensuring that any previously created instances had been pulled out of the pool and disposed of properly.  So for this PR, we could just leave the implementation of `.clear` as is.  I don't think the implementation as it stands on the branch currently would be a regression in these cases.  The next attempt to borrow from the pool after either of the code paths above are executed would still return a poison pill to the caller.